core/store/src/flat/store_helper.rs

```diff
@@ -36,40 +39,50 @@ impl FlatStateColumn {
         #[cfg(feature = "protocol_feature_flat_state")]
         match self {
             FlatStateColumn::State => crate::DBCol::FlatState,
-            FlatStateColumn::Deltas => crate::DBCol::FlatStateDeltas,
+            FlatStateColumn::Changes => crate::DBCol::FlatStateChanges,
+            FlatStateColumn::DeltaMetadata => crate::DBCol::FlatStateDeltaMetadata,
             FlatStateColumn::Misc => crate::DBCol::FlatStateMisc,
         }
         #[cfg(not(feature = "protocol_feature_flat_state"))]
         panic!("protocol_feature_flat_state feature is not enabled")
     }
 }
 
-pub fn get_delta(
+pub fn get_delta_changes(
     store: &Store,
     shard_id: ShardId,
     block_hash: CryptoHash,
-) -> Result<Option<FlatStateDelta>, FlatStorageError> {
+) -> Result<Option<FlatStateChanges>, FlatStorageError> {
     let key = KeyForFlatStateDelta { shard_id, block_hash };
     Ok(store
-        .get_ser::<FlatStateDelta>(FlatStateColumn::Deltas.to_db_col(), &key.try_to_vec().unwrap())
+        .get_ser::<FlatStateChanges>(
+            FlatStateColumn::Changes.to_db_col(),
+            &key.try_to_vec().unwrap(),
+        )
         .map_err(|_| FlatStorageError::StorageInternalError)?)
 }
 
 pub fn set_delta(
     store_update: &mut StoreUpdate,
     shard_id: ShardId,
-    block_hash: CryptoHash,
     delta: &FlatStateDelta,
 ) -> Result<(), FlatStorageError> {
-    let key = KeyForFlatStateDelta { shard_id, block_hash };
+    let key = KeyForFlatStateDelta { shard_id, block_hash: delta.metadata.block.hash }
+        .try_to_vec()
+        .map_err(|_| FlatStorageError::StorageInternalError)?;
     store_update
-        .set_ser(FlatStateColumn::Deltas.to_db_col(), &key.try_to_vec().unwrap(), delta)
-        .map_err(|_| FlatStorageError::StorageInternalError)
+        .set_ser(FlatStateColumn::Changes.to_db_col(), &key, &delta.changes)
+        .map_err(|_| FlatStorageError::StorageInternalError)?;
+    store_update
```

Comment:
It's a bit confusing that we set/remove metadata, but never read it. Maybe also implement `get_delta` here and call it on flat storage creation, and get rid of `ChainAccessForFlatStorage` occurrences later?

By this I'm trying to ensure that we write correct metadata even on intermediate branches, so we can safely run flat storage experiments.

Side thought: now it's a bit more concerning that we never read changes from disk on real-time load. Maybe we need to add a test later which limits capacity of cache with changes, or verifies consistency of cache with DB.

Reply:
Updated the code to use metadata for `FlatStorage` creation.

Reply:
good point, I've added that to the description